### PR TITLE
Gossip enhance prework

### PR DESCRIPTION
**Pull request by rustyrussell**
I've been working on trying to test the gossip enhancements proposed, and it requires some more infrastructure around features and gossip messages.

In particular, we'll be relaying more messages for gossipd, so that needs to be enhanced: we've currently been ignoring errors when we should have been closing connections according to the spec.

### CODE REVIEW

File: channeld/channel.c [hunk 1]
```diff
@@ -306,12 +306,17 @@ static void enqueue_peer_msg(struct peer *peer, const u8 *msg TAKES)
...
+	u64 gossip_index;
 
-	if (!fromwire_gossip_send_gossip(msg, msg, &peer->gossip_index, &gossip))
+	if (!fromwire_gossip_send_gossip(msg, msg, &gossip_index, &gossip))
 		status_failed(STATUS_FAIL_GOSSIP_IO,
 			      "Got bad message from gossipd: %s",
 			      tal_hex(msg, msg));
 
+	/* Zero is a special index meaning this is unindexed gossip. */
+	if (gossip_index != 0)
```
Comment (cdecker) [inline channeld/channel.c]: I assume this is used for the range signalling for a later commit?

File: gossipd/gossip.c [hunk 2]
```diff
@@ -291,6 +291,22 @@ static void reached_peer(struct daemon *daemon, const struct pubkey *id,
...
 }
 
+static void queue_peer_msg(struct peer *peer, const u8 *msg TAKES)
+{
+	const u8 *send;
+
+	if (peer->local) {
+		msg_enqueue(&peer->local->peer_out, msg);
+		return;
+	}
```
Comment (cdecker) [inline gossipd/gossip.c]: I try to personally avoid tearing apart an if-else statement into if-return, especially when both branches are equally likely. It just helps with readability.

File: gossipd/routing.c [hunk 3]
```diff
@@ -566,8 +611,12 @@ const struct short_channel_id *handle_channel_announcement(
 					   &pending->node_id_2,
 					   &pending->bitcoin_key_1,
 					   &pending->bitcoin_key_2)) {
+		err = towire_errorfmt(rstate, NULL,
+				      "Malformed channel_announcement %s",
+				      tal_hex(pending, pending->announce));
 		tal_free(pending);
-		return NULL;
+		*scid = NULL;
```
Comment (cdecker) [inline gossipd/routing.c]: Having these null-sets plastered all over the place seems dangerous, wouldn't it not be better to just set NULL once on top? I know you like `valgrind` telling you about unset things, but there is no extra information to be had here, other than "oops forgot to set".

Comment (cdecker) [reply gossipd/routing.c]: I'd even go as far and say the caller should set whatever special value he is expecting for unset.

File: gossipd/gossip.c [hunk 4]
```diff
@@ -1369,7 +1371,11 @@ static void gossip_send_keepalive_update(struct routing_state *rstate,
 	status_trace("Sending keepalive channel_update for %s",
 		     type_to_string(tmpctx, struct short_channel_id, &scid));
 
-	handle_channel_update(rstate, update);
+	err = handle_channel_update(rstate, update);
```
Comment (cdecker) [inline gossipd/gossip.c]: This should probably be fatal, it's an update we generate so if we don't accept it, that's on us.

File: gossipd/gossip.c [hunk 5]
```diff
@@ -1927,7 +1934,11 @@ static struct io_plan *handle_disable_channel(struct io_conn *conn,
 			      strerror(errno));
 	}
 
-	handle_channel_update(daemon->rstate, msg);
+	err = handle_channel_update(daemon->rstate, msg);
```
Comment (cdecker) [inline gossipd/gossip.c]: Same here, if we can't be trusted to disable tell ourselves about a channel we disable, then that's on us.

File: gossipd/routing.c [hunk 6]
```diff
@@ -793,10 +805,11 @@ bool handle_pending_cannouncement(struct routing_state *rstate,
 		pubkey_eq(&pending->node_id_2, &rstate->local_id);
 
 	/* Did we have an update waiting?  If so, apply now. */
+	/* FIXME: We don't remember who sent us updates, so we can't error them */
 	if (pending->updates[0])
-		handle_channel_update(rstate, pending->updates[0]);
+		tal_free(handle_channel_update(rstate, pending->updates[0]));
```
Comment (cdecker) [inline gossipd/routing.c]: We should log them maybe?

**Comment by cdecker** [comment]
@ZmnSCPxj `openingd` and `closingd` are very constrained as to what messages are allowed during the opening and closing flows, so they do not handle gossip messages since that'd interfere with the flows. They are shortlived daemons so there is little use in plumbing in the gossip machinery.

**Comment by ZmnSCPxj** [comment]
@cdecker my poor knowledge is that nothing in particular prevents a peer from sending gossip messages during opening or closing.  For example if I make a custom client that performs `open_channel` but then forgets the flow, ignores `accept_channel`, and never sends `funding_created`, but starts sending gossip messages, then that is "valid" under the protocol, especially if my custom client does not particularly want an open channel anyway.  Or is there something in the BOLT spec which prevents gossip during opening and closing flows?

**Comment by cdecker** [comment]
You are of course right, but we currently just drop them I think.

**Comment by rustyrussell** [comment]
Both those daemons will relay gossip they receive from the peer, *but* never relay from gossipd to the peer.  ie. no outgoing gossip.

File: gossipd/gossip.c (hunk 4 above)
Comment (rustyrussell) [reply gossipd/gossip.c]: I'm generally trying to cut down on fatal errors, especially now we have status_broken().  But perhaps I'm premature: will change to fatal().

File: gossipd/routing.c (hunk 3 above)
Comment (rustyrussell) [reply gossipd/routing.c]: I was aiming for minimal change, but I think you're right.  Making the caller set it is a bit too far: this is effectively a trinary return: not found, found, malformed.  My instincts are to try to split the malformed test off into a separate function, but that's awkward here, hence the weirdness.

File: gossipd/gossip.c (hunk 2 above)
Comment (rustyrussell) [reply gossipd/gossip.c]: Agreed.

File: gossipd/routing.c (hunk 3 above)
Comment (rustyrussell) [reply gossipd/routing.c]: Actually, I turned them into gotos, which is even clearer IMHO.

File: gossipd/gossip.c (hunk 4 above)
Comment (rustyrussell) [reply gossipd/gossip.c]: Actually, it *is* fatal (when gossipd dies, main daemon dies):

```
	if (err)
		status_failed(STATUS_FAIL_INTERNAL_ERROR,
			      "rejected keepalive channel_update: %s",
			      tal_hex(trc, err));
```

File: gossipd/routing.c (hunk 6 above)
Comment (rustyrussell) [reply gossipd/routing.c]: Yes, good idea.

**Comment by cdecker** [comment]
Excellent work :+1:

ACK f0fc165

**Pull request merged**